File: src/lock/lock_type.py
"""
Lock Type Enum - Tipe-tipe lock dalam sistem
"""

from enum import Enum


class LockType(Enum):
    """
    Enum untuk tipe lock dalam distributed locking
    """
    
    READ = "READ"
    """Read lock (shared) - multiple readers allowed"""
    
    WRITE = "WRITE"
    """Write lock (exclusive) - only one writer allowed"""
    
    def is_compatible(self, other: 'LockType') -> bool:
        """
        Cek apakah lock type ini kompatibel dengan lock type lain
        
        Args:
            other: Lock type lain
            
        Returns:
            True jika kompatibel
        """
        if self == LockType.READ and other == LockType.READ:
            return True
        return False

File: src/main/java/com/simulator/lock/LockType.java
package com.simulator.lock;

/**
 * Jenis-jenis lock dalam sistem distributed
 */
public enum LockType {
    SHARED("Shared Lock"),
    EXCLUSIVE("Exclusive Lock"),
    INTENT_SHARED("Intent Shared"),
    INTENT_EXCLUSIVE("Intent Exclusive");
    
    private final String description;
    
    LockType(String description) {
        this.description = description;
    }
    
    public String getDescription() {
        return description;
    }
    
    public boolean isShared() {
        return this == SHARED || this == INTENT_SHARED;
    }
    
    public boolean isExclusive() {
        return this == EXCLUSIVE || this == INTENT_EXCLUSIVE;
    }
    
    @Override
    public String toString() {
        return description;
    }
}